Add Breakpad for crash dump generation

hhyyrylainen opened the pull request:
<!--
Pull requests should always be made for the `master` branch first, as that's
where development happens and the source of all future stable release branches.

Relevant fixes are cherry-picked for stable branches as needed.

Do not create a pull request for stable branches unless the change is already
available in the `master` branch and it cannot be easily cherry-picked.
Alternatively, if the change is only relevant for that branch (e.g. rendering
fixes for the 3.2 branch).
-->

I started working on a simple Breakpad crash reporting integration to Godot related to this proposal: https://github.com/godotengine/godot-proposals/issues/1896

Some current caveats:
- ~~I have not tested this with Godot 4, I originally made this against the 3.4 stable label and cherry picked to this branch (that other branch is here: https://github.com/Revolutionary-Games/godot/tree/crash_dumper_3.x). I only did a very quick compile fix and checked that Godot starts on this branch. There didn't seem to be a working sample project available in the asset library(?) so I couldn't test with a project.~~
- ~~There's some cleanup left to do regarding which files to include and which need to be compiled~~
- ~~My local install of clang-format doesn't work with the options in this repo so I couldn't format the files with it (clang-format failed with code 1: /home/hhyyrylainen/Projects/godot/.clang-format:151:1: error: unknown key 'SpacesInLineCommentPrefix')~~ update to Fedora 35 fixed this by having a newer clang version
- ~~I noticed that Windows crash reporter on destruction doesn't disable itself. seems like nothing else either disables it on shutdown. Is that intended? It's not consistent with Linux. For my own testing I added the line to disable it in the destructor, but I didn't commit that here~~ I added this code
- ~~I haven't yet tested whether the created crashdumps can be decoded with stackwalk (I'm going to work on my build scripts next to make this work, I'm not confident enough with the Godot build process yet to make a PR to the build scripts repo). Also Windows is less tested than Linux~~ Windows crash dumps only work with the special mingw supporting stackwalk, but no special version is required in the Godot repo when building Godot, so this is just a minor inconvenience.
- ~~I don't have a mac to develop on, so~~ support for that needs to be done later

bruvzg commented:
> Edit: I meant to open this as a draft, and now there doesn't seem to be a button to convert this into a draft...

![Screenshot 2021-12-17 at 11 58 24](https://user-images.githubusercontent.com/7645683/146526481-ce29450a-82f3-46c0-bdca-e975fa9d80d4.png)

hhyyrylainen commented:
Oh yeah, thanks. I misremembered where that button was supposed to be, and couldn't see it there...

fire commented on modules/breakpad/SCsub:
```diff
@@ -0,0 +1,196 @@
+
+breakpad_src = [
+    "src/common/convert_UTF.cc",
+    # "src/common/long_string_dictionary.cc",
```
We tend to remove commented out code that isn't used, unless you want to signal that is is unused as a warning.

hhyyrylainen replied:
I'll remove (most) of these commented out things once I'm done. I didn't remove these yet as I'm not sure if these are required for other platforms (I only tested compiling on Linux removing these files from the compile and putting them back in until all linking errors were gone).

fire commented on modules/breakpad/SCsub:
```diff
@@ -0,0 +1,196 @@
+        "src/client/ios/handler/ios_exception_minidump_generator.mm",
+    ]
+
+# if solaris:
```
Are you using Godot on Illumos / Solaris? :D Cool!

hhyyrylainen replied:
Well there are files for solaris in Breakpad, so I thought it would be at least theoretically useful to keep the list of files for solaris here.

fire commented on modules/breakpad/breakpad_linuxbsd_windows.cpp:
```diff
@@ -0,0 +1,130 @@
+        return;
+    }
+
+#ifdef _WIN32
```
Probably want to use PLATFORM_WINDOWS, but it's not a big thing.

hhyyrylainen replied:
PLATFORM_WINDOWS doesn't seem to be a thing? That's why I selected "_WIN32" as the thing to check for.

```
[hhyyrylainen@dakara godot]$ ag PLATFORM_WINDOWS
thirdparty/doctest/doctest.h
349:#define DOCTEST_PLATFORM_WINDOWS
2841:#ifdef DOCTEST_PLATFORM_WINDOWS
2859:#else // DOCTEST_PLATFORM_WINDOWS
2864:#endif // DOCTEST_PLATFORM_WINDOWS
2995:#if defined(DOCTEST_PLATFORM_WINDOWS)
2997:#else // DOCTEST_PLATFORM_WINDOWS
3000:#endif // DOCTEST_PLATFORM_WINDOWS
3007:#elif defined(DOCTEST_PLATFORM_WINDOWS)
3018:#else  // DOCTEST_PLATFORM_WINDOWS
3024:#endif // DOCTEST_PLATFORM_WINDOWS
3618:#ifdef DOCTEST_PLATFORM_WINDOWS
4186:#ifdef DOCTEST_PLATFORM_WINDOWS
4334:#else // DOCTEST_PLATFORM_WINDOWS
4413:#endif // DOCTEST_PLATFORM_WINDOWS
4421:#ifdef DOCTEST_PLATFORM_WINDOWS
5075:#ifdef DOCTEST_PLATFORM_WINDOWS
5078:#endif // DOCTEST_PLATFORM_WINDOWS
5284:#ifdef DOCTEST_PLATFORM_WINDOWS
5286:#else // DOCTEST_PLATFORM_WINDOWS
5288:#endif // DOCTEST_PLATFORM_WINDOWS
5373:#ifdef DOCTEST_PLATFORM_WINDOWS
5376:#endif // DOCTEST_PLATFORM_WINDOWS
5956:#ifdef DOCTEST_PLATFORM_WINDOWS
5990:#endif // DOCTEST_PLATFORM_WINDOWS
6342:#ifdef DOCTEST_PLATFORM_WINDOWS
6345:#endif // DOCTEST_PLATFORM_WINDOWS
```

versus:
```
[hhyyrylainen@dakara godot]$ ag _WIN32
core/math/static_raycaster.h
38:#if defined(_WIN32) && defined(_MSC_VER)

core/typedefs.h
75:#ifdef _WIN32

drivers/vulkan/SCsub
25:    env.AppendUnique(CPPDEFINES=["VK_USE_PLATFORM_WIN32_KHR"])

modules/gdnative/include/gdnative/gdnative.h
38:#if defined(_WIN32) || defined(__ANDROID__)
50:#else // !_WIN32 && !__APPLE__
58:#elif defined(_WIN32)

modules/raycast/godot_update_embree.py
239:#elif defined(_WIN32)

modules/breakpad/breakpad_linuxbsd_windows.cpp
36:#ifdef _WIN32
44:#ifdef _WIN32
62:#ifdef _WIN32
103:#ifdef _WIN32
116:#ifndef _WIN32
127:#ifdef _WIN32

platform/uwp/detect.py
161:    env.Append(CPPDEFINES=[("WINVER", winver), ("_WIN32_WINNT", winver), "WIN32"])

-- and there's a ton more matches after this...
```

fire replied:
Edited: `WINDOWS_ENABLED`

hhyyrylainen replied:
Yeah, I can change to that.

hhyyrylainen replied:
Made that change.

fire commented:
I have approved cicd.

bruvzg commented on modules/breakpad/SCsub:
```diff
@@ -0,0 +1,196 @@
+        "src/client/mac/crash_generation/InspectorMain.mm",
+        "src/client/mac/crash_generation/crash_generation_client.cc",
+        "src/client/mac/crash_generation/crash_generation_server.cc",
+        "src/client/mac/gcov/libgcov.a",
+        "src/client/mac/handler/breakpad_nlist_64.cc",
+        "src/client/mac/handler/dynamic_images.cc",
+        "src/client/mac/handler/exception_handler.cc",
+        "src/client/mac/handler/minidump_generator.cc",
+        "src/client/mac/handler/obj-cTestCases-Info.plist",
+        "src/client/mac/handler/protected_memory_allocator.cc",
+        "src/common/mac/Breakpad.xcconfig",
+        "src/common/mac/BreakpadDebug.xcconfig",
+        "src/common/mac/BreakpadRelease.xcconfig",
+        "src/common/mac/GTMLogger.m",
+        "src/common/mac/HTTPGetRequest.m",
+        "src/common/mac/HTTPPutRequest.m",
+        "src/common/mac/HTTPRequest.m",
+        "src/common/mac/HTTPSimplePostRequest.m",
+        "src/common/mac/MachIPC.mm",
+        "src/common/mac/arch_utilities.cc",
+        "src/common/mac/bootstrap_compat.cc",
+        "src/common/mac/encoding_util.m",
+        "src/common/mac/file_id.cc",
+        "src/common/mac/launch_reporter.cc",
+        "src/common/mac/macho_id.cc",
+        "src/common/mac/macho_reader.cc",
+        "src/common/mac/macho_utilities.cc",
+        "src/common/mac/macho_walker.cc",
+        "src/common/mac/string_utilities.cc",
+    ]
+
+if env["platform"] == "iphone":
+    breakpad_src += [
+        "src/client/ios/Breakpad.mm",
+        "src/client/ios/Breakpad.xcodeproj/project.pbxproj",
+        "src/client/ios/BreakpadController.mm",
+        "src/client/ios/Breakpad_Prefix.pch",
+        "src/client/ios/exception_handler_no_mach.cc",
+        "src/client/ios/handler/ios_exception_minidump_generator.mm",
+    ]
```
`.pbxproj`, `.xcconfig`, `.pch`, `.plist` and `.a` are not source files, and should not be added here.

Probably `src/client/mac/Framework/*.mm` are intended to be built as a separate framework.

hhyyrylainen replied:
Removed those files. Also deleted the `.a` file as I couldn't find any reference to it. So probably better to remove it unless it turns out it is really needed on mac.

As I said I don't have a mac to actually build on, but it seems that building breakpad is a bit of a mess on mac. And maybe that separate framework is not even needed at all? https://stackoverflow.com/questions/5404636/building-google-breakpad-on-mac-os-x

fire commented:
Some more static check failures and try to collapse the commits by squashing. We prefer one unless you have a logical reason (we allow to have a few commits).

hhyyrylainen commented:
I have ran clang-format and renamed that one place where a file was incorrectly referred to in a file copyright header. I haven't pushed those changes yet as I wanted to check the Windows compile and determine if some of the common files are even needed for any platform. I'll do a manual squash as the last thing once I don't need to tweak the code anymore, as that'll get rid of some intermediate files that would otherwise be in git history.

hhyyrylainen commented:
All issues detected by CI should now hopefully be fixed. I also removed the commented out parts in the build config file and did some last tweaks. Also made sure it compiles on Windows, I'm not super confident about how the string conversion should be used in 4.0, in 3.4 I think I was able to figure out it pretty well.

I *think* this now has all the changes on the engine side that are needed for crash dump generation. Though, I've not yet gotten to test the entire flow of creating a crash dump and being able to decode it. As an additional difficulty Windows MinGW builds seem a bit problematic regarding extracting symbols but this breakpad fork: https://github.com/DaemonEngine/breakpad seems to have a working tool for that.

hhyyrylainen commented:
I've just ran into a pretty big issue, it seems even though I found a tool that can dump the symbols of a mingw created exe on Linux, it doesn't seem to contain any symbols besides the standard library. It doesn't even detect that any of the Godot source files.
I tried to use a modified build script with `LINKFLAGS=-Wl,--build-id=md5 CCFLAGS=-g CFLAGS=-g CXXFLAGS=-g` however that didn't either result in any more found symbols. Is there some option I'm missing to get Windows debug symbols for Godot?
I don't want to fallback to compiling the Windows version on Windows as that basically requires Windows 10 to run correctly (though I suppose officially Windows 8 is only getting one more year of updates from Microsoft), and I have not found a way to make it work. According to all Microsoft documentation the universal C runtime that visual studio 2019 uses, should be installable on older Windows versions, but I haven't seen anyone able to get that working...

fire commented:
Let me get back to you but I think I posted a llvm-mingw workflow.

fire commented:
Here's the exact command I use.

```bash
PATH=/opt/llvm-mingw/bin:$PATH scons werror=no platform=windows target=release_debug -j`nproc` use_lto=no deprecated=no use_mingw=yes use_llvm=yes use_thinlto=yes LINKFLAGS=-Wl,-pdb= CCFLAGS='-g -gcodeview' debug_symbols=no"
```

The import part is llvm-mingw and `LINKFLAGS=-Wl,-pdb= CCFLAGS='-g -gcodeview' `

hhyyrylainen commented:
Kind of expected, but the Godot podman build image scripts don't install llvm, so I get this error with that:
```
sh: line 1: x86_64-w64-mingw32-clang++: command not found
```
it shouldn't be too difficult to modify my local copy to also install llvm in the build image to test out that way of making the build. First I'll try that thinlto one with the mingw gcc to see if that has any effect.

fire commented on SConstruct:
```diff
@@ -135,6 +135,7 @@ opts.Add(BoolVariable("opengl3", "Enable the OpenGL/GLES3 video driver", True))
 opts.Add("custom_modules", "A list of comma-separated directory paths containing custom modules to build.", "")
 opts.Add(BoolVariable("custom_modules_recursive", "Detect custom modules recursively for each specified path.", True))
 opts.Add(BoolVariable("use_volk", "Use the volk library to load the Vulkan loader dynamically", True))
+opts.Add(BoolVariable("breakpad_enabled", "Enable Breakpad crash dump creation.", False))
```
The common pattern is "use_breakpad". Thoughts?

hhyyrylainen replied:
I can change it easily enough.

fire replied:
Can you change this to to "use_breakpad" unless you have reasoning against?

hhyyrylainen replied:
Yes, I can. I got stuck yesterday trying to fix that problem I commented about (https://github.com/godotengine/godot/pull/56014#issuecomment-998253060), and I didn't want to switch over to changing this as I have a bunch of experimental local changes.

hhyyrylainen replied:
I renamed the scons option and the defines used in C++ as volk seemed to also use that format of preprocessor defines.

hhyyrylainen commented:
I ran into another issue: in release mode on Linux when Godot doesn't register the signal handlers, but instead I leave that to Breakpad, it doesn't work for some reason. I might need to do a workaround where the Godot handlers are always installed, or perhaps the breakpad version I got has a bug, but that would be a pretty serious slip by Google...

fire commented:
How do I recreate your bug case? Maybe posting some info here can reveal the answer as I'm not in this area.

hhyyrylainen commented:
After debugging with gdb I think I figured it out. The mono runtime actually messes with the signals. I noticed when putting breakpoints on `signal` and `sigaction` that the mono initialization actually overrides the handlers that breakpad installs, however for some reason those signal handlers from mono don't mess with signal handlers installed using the `signal` function so only breakpad is impacted as it uses the more extensive `sigaction` function.

Here's where mono messes with the signal handlers:
```
(gdb) bt
#0  __GI___sigaction (sig=35, act=0x0, oact=0x7fffffffc2d0) at sigaction.c:26
#1  0x0000000000cdcd6d in mono_threads_suspend_search_alternative_signal ()
#2  0x0000000000cdcf65 in mono_threads_suspend_init_signals ()
#3  0x0000000000add5ba in mini_init ()
#4  0x0000000000a353eb in mono_jit_init_version ()
#5  0x0000000000d80bb6 in (anonymous namespace)::gd_initialize_mono_runtime ()
    at modules/mono/mono_gd/gd_mono.cpp:199
#6  GDMono::initialize (this=0x4ce3bf0) at modules/mono/mono_gd/gd_mono.cpp:378
#7  0x0000000001088d1a in CSharpLanguage::init (this=0x4257ec0)
    at modules/mono/csharp_script.cpp:113
#8  0x000000000250b7df in ScriptServer::init_languages ()
    at core/script_language.cpp:170
#9  0x0000000000d4f4bb in Main::setup2 (p_main_tid_override=<optimized out>)
    at main/main.cpp:1489
#10 0x0000000000d547a3 in Main::setup (execpath=<optimized out>, 
    argc=<optimized out>, argv=<optimized out>, p_second_phase=<optimized out>)
    at main/main.cpp:1232
#11 0x00000000009f8469 in main (argc=1, argv=0x7fffffffd9f8)
    at platform/x11/godot_x11.cpp:48
```

This means that using a workaround of always registering the Godot crash handler signal handlers and passing that onto breakpad, should work. I don't have enough time to test that today, though.

hhyyrylainen commented:
Seems like not even adding code like:
```
#elif defined(BREAKPAD_ENABLED)

	signal(SIGSEGV, handle_crash);
	signal(SIGFPE, handle_crash);
	signal(SIGILL, handle_crash);

	initialize_breakpad(false);
#endif
```
works, even though that's the same as in debug mode. Note that the editor properly runs the crash handlers.

So there's definitely something in the Linux Mono runtime when running in release mode exported game that makes it override Godot's and also Breakpad's signal handlers. I doubt that even Godot's own crash handler was able to work in this mode, but then again it's always disabled in the release mode, so I guess no one has encountered this before.

It doesn't seem there's an easily tweakable place in gd_mono.cpp where this happens. So I might just have to put in a workaround in gd_mono.cpp (or some place higher up in the callstack), where I re-initialize Breakpad (on Linux) just after Mono has been loaded to fix the signals it messed up.

fire commented:
Happy Holidays. Do you think it's ok to drop mono support with as a platform feature? Like use the feature plans to add per platform support for breakpad crashdump generation. It's an option if you ran into trouble.

hhyyrylainen commented:
As far as I can tell, only Linux experiences this issue. I've been trying to fix the Linux side so I only tested with Wine, but there the crash dumped log was printed fine before the Wine crash reporter opened. So I'm fairly confident the Windows side of things works.
I'm (almost) ready to give up and only have crash dumps working with Mono on Windows, which basically 100% covers all reports of Godot engine not running on player's computers that I've received. On Linux as a workaround I can try to ask players to send me full (compressed) core files.

hhyyrylainen commented:
Actually, it might be just that *shutting down* the process causes the signal handlers to be unregistered. I just tested in gdb by pausing the running game and giving it the SIGSEGV signal, which does result in the crash handler running. I'll need to investigate a bit more (especially where I could put in my manual change to cause a consistent crash) regarding this. But I might be able to wrap that up today and do the few remaining tweaks to this PR.

hhyyrylainen commented:
Okay, I was a bit optimistic, but that got me on the right track. The issue was that I had put some crashing code into the crash handler disable method, which never ran on Linux because it seems that mono also messes with signal handlers when it is shutting down. Moving that crashing code to the linux main loop, I got things working. I think I've also now addressed all the comments that came up in code reviews so far.

I tried quite many combinations but in the end the result is that the Linux signal handlers have to be re-installed just after mono is initialized, that way they work correctly.
![Kuvakaappaus - 2021-12-28 21-02-38](https://user-images.githubusercontent.com/3796411/147605638-d5df01a4-47d3-4ab0-8fec-f45ec0ecf6b5.png)



Note: that I have only tested this with 3.4 branch, but I hope everything I've done still applies to 4.0. If there is an example project for Godot 4.0, I can try that, in case there is something that is different there. For my own use, I'll only for now really need this functionality on the 3.4 branch.

fire commented:
What happen in godot 4 when you make a sample project that calls the 3.0 crash function?

hhyyrylainen commented:
Seems like I can't built the non-editor binary of 4.0 at all:
```
[ 73%] Linking Program        ==> bin/godot.linuxbsd.opt.64
/usr/bin/ld: modules/text_server_adv/libicu_builtin.linuxbsd.opt.64.a(stringtriebuilder.linuxbsd.opt.64.o): warning: relocation against `__cxa_pure_virtual' in read-only section `.rodata._ZTVN6icu_7017StringTrieBuilder15LinearMatchNodeE[_ZTVN6icu_7017StringTrieBuilder15LinearMatchNodeE]'
/usr/bin/ld: warning: creating DT_TEXTREL in a PDE
[ 73%] scons: done building targets.
[Time elapsed: 00:03:21.887]
```
so I can't currently test it at all, but I tested earlier that the editor does create a crash dump even on the 4.0 branch. And I just verified that it still does work.

hhyyrylainen commented:
Will this be reviewed again soon? I'm planning on releasing a game version that would have this in it in a few weeks. I'll use the 3.4 branch as a base but it would be a bit cleaner for me to make the custom branch if this was already merged into master and cherry-picked to the 3.x versions.

mhilbrunner commented:
> Will this be reviewed again soon? I'm planning on releasing a game version that would have this in it in a few weeks. I'll use the 3.4 branch as a base but it would be a bit cleaner for me to make the custom branch if this was already merged into master and cherry-picked to the 3.x versions.

Perfectly understandable, but it is currently unlikely to be cherry-picked into 3.x in the next few weeks as the core team is pretty busy with releasing Godot 4.0 alpha 1, and a change such as this will take time and multiple reviews/approvals by different people as it is pretty core. I'll try to review it soon, but you should assume this will take some time to get discussed/reviewed. (We also have quite some backlog of PRs to get down.)

I apologize for the inconvenience :) Hope your release goes well!

hhyyrylainen commented:
Thanks.
I assumed as much not getting a response for a few days. I already made the 3.4 based branch, which didn't have too many merge conflicts to solve: https://github.com/Revolutionary-Games/godot/tree/crash_dumper_3.4.2
Which builds fine now that I found a workaround for https://github.com/godotengine/build-containers/issues/101 also looks like the breakpad dumper version I found doesn't work with that newer mingw version, which I still need to solve. But that isn't a problem with the code that is in this PR.

hhyyrylainen commented:
It shows 4 files for me?
```
COPYRIGHT.txt
main/main.cpp
modules/mono/csharp_script.cpp
thirdparty/README.md 
```

I can fix the merge conflicts if this has a chance to get merged once I do, I don't want to end up fixing merge conflicts over and over...

fire commented:
Well, due to the lack of proposal consensus, I wanted to try it on my branch and see if it does what it says it does.

Wanted to ask for a rebase before trying it myself.

We have a lot of crashes in my game and I wanted to try it so I can have a more effective proposal.

hhyyrylainen commented:
I have now fixed the merge conflicts.

fire commented:
Doing some initial review. The musl folder doesn't seem to be used and maybe others.

hhyyrylainen commented:
I won't say I 100% remember, but I'm at least somewhat sure there was some build error without that folder...

hhyyrylainen commented:
I removed the musl folder and the one elf_reader file which seemed to be the only thing depending on it. Kind of looks like that was only necessary in the breakpad dumper binaries so the final game executable shouldn't need it.
Also I rebased this onto the latest master now.

hhyyrylainen commented:
With 4.0 out now, is there a chance that this could get merged at some point? If there's a chance I'll work through the merge conflicts and test everything again to check that this still works. I also now have a development mac mini so I could also try to get the mac side of things working if that means there's a better chance for this to be merged.

fire commented:
The 4.1 feature freezed but I will try my best to get the support from the development team.

As I want to try breakpad, do you know how difficult it is to have a master branch and if 4.0 is cherry pickable from it.

hhyyrylainen commented:
I suspect this is currently at least slightly broken for 4.x as this was written for Mono. Other than updating the hooks a bit and testing that dotnet runtime doesn't destroy the signal listeners, this shouldn't be too difficult to update. 
If it does destroy the listeners then a similar workaround as for mono is needed. For mono re-registering the handlers after after mono was initialized worked fine.

ed-megacrit committed:
Added Breakpad crash dump generation

Fixed compiling on Windows

Make crash dump printing work better on Windows

Tweaked Breakpad build

Fixed Breakpad build for Godot 4

Switched to WINDOWS_ENABLED

Removed some non buildable files listed in the files to build

Ran clang format

Fixed the other formatting issues detected by CI

Removed a comment and added clarifying comment on crash dump message

as to why it is printed twice on Windows

Make an ugly string conversion to make Windows build work

Tweaked the build configuration and formatted again

removed lss

Add lss properly

Reinitialize breakpad after mono initialization on Linux

otherwise the breakpad signal handlers are not active

Disable Windows crash handler on destruction similarly to Linux

Renamed breakpad_enabled to use_breakpad

Forgot to wrap one piece of code inside ifdef USE_BREAKPAD

Updated copyright years in the added files

Fix register types for breakpad

Fixed dir access

Removed musl and elf_reader

which was the only thing seemingly depending on it

Updated header guards

Removed the memdelete call

ed-megacrit committed:
Fixups

ed-megacrit committed:
Remove mono re-registration, replace accidentally removed line

Chaosed0 commented:
I rebased this onto the current master branch and submitted a PR to the fork: https://github.com/Revolutionary-Games/godot/pull/1

I would really like to see this get in, as there's currently no way to obtain postmortem debugging information for crashes that occur in the wild.

fire commented:
Tempted to check how much work to get this to 4.3-beta1

ed-megacrit committed:
Update breakpad

ed-megacrit committed:
Delete unused files

hhyyrylainen committed:
Added missing breakpad source file compile for Linux

hhyyrylainen committed:
Try to fix .gitignore matching Breakpad source files

hhyyrylainen commented:
Thanks to @Chaosed0 this PR has been updated to be rebased on master and should be working in Godot 4. I haven't personally done a full end to end test as my game currently always crashes on shutdown (due to: https://github.com/godotengine/godot/issues/89941) so this wouldn't be helpful to me currently in any case. 

Also I didn't yet squash the commits into one as this probably has some tweaks left to do and I was unsure how I should squash things in the case of having multiple authors on a PR.

hhyyrylainen commented on modules/breakpad/breakpad.h:
```diff
@@ -0,0 +1,48 @@
+void report_user_data_dir_usable();
+
+// Due to Mono runtime initialization in release mode overriding signal handlers, Breakpad needs re-initialization after loading it
+void report_mono_loaded_to_breakpad();
```
Looks like there are some leftover mono bits of the code. I can remove these bits when someone would be able to do a PR review / there's a chance that this PR could be betting any closer to being merged.

hhyyrylainen commented on thirdparty/README.md:
```diff
@@ -71,6 +71,26 @@ Applied upstream PR https://github.com/BinomialLLC/basis_universal/pull/344 to
+## breakpad
+
+- Upstream: https://chromium.googlesource.com/breakpad/breakpad/
+- Version: git (605c51ed96ad44b34c457bbca320e74e194c317e, 2021)
```
This is now outdated. @Chaosed0 could I get the commit hash you used to update the Breakpad files so this could be changed to be accurate?

Chaosed0 replied:
Yes, here it is: `527331a471207d28d015e210c12fb8bb8b6732fd`

hhyyrylainen committed:
Update the version hash breakpad is said to be used from

hhyyrylainen committed:
Made copyright header changes that the checks wanted

huwpascoe commented:
Please could you squash those commits?

hhyyrylainen commented:
What's the policy on squashing commits from multiple authors? Adding co-authored by lines to the squashed commit is required to not lose attribution for the multiple people's work.

Also I didn't squash this yet as it seems unlikely that this would be any more likely to be merged sooner even if I did that. And I kind of suspect that this is going to break at least once more before this can be merged...

I can squash this up if it makes it any more likely that this PR would move forward (and I'd need to know how to handle squashing commits from multiple authors). When is the next possible merge window opening up?

huwpascoe commented:
Not sure, but 4.3 is nearly out so this is probably the best time to get it ready to guarantee in 4.4.

Considering how many issues there are with completely worthless crash messages, this PR should be considered top priority, at least I think so. I've been told by @akien-mga that testing PRs are more likely to get them merged. So this can be the pilot PR to prove it.

> Please include this when reporting the bug to the project developer.
[1] error(-1): no debug info in PE/COFF executable

<sub>programmer humor</sub>

### Checkout and Build
- [x] pass (but with warnings)
Rebased onto master (383a6e4ef285fddcb0e1cbe28a20332c7e5a815b) without conflicts.
Windows 10, x64, MSVC `use_breakpad = "yes"`
```
warning STL4043: stdext::checked_array_iterator, stdext::unchecked_array_iterator, and related factory functions are non-Standard extensions and will be removed in the future. std::span (since C++20) and gsl::span can be used instead.
You can define _SILENCE_STDEXT_ARR_ITERS_DEPRECATION_WARNING or _SILENCE_ALL_MS_EXT_DEPRECATION_WARNINGS to suppress this warning.
```
### Test
I placed a bug into a known script-callable function.
```
#pragma warning(push, 0)
	int *bug = nullptr;
	*bug = 123;
#pragma warning(pop, 0)
```
After building I removed the pdb to ensure a worthless log.
Tested with both the editor and the running app.
### Result
- [x] pass

In addition to the usual log, a .dmp file was created in "%appdata%\Godot\app_userdata\testproj\crashes"
Opening the .dmp with visual studio pointed to source file and the bug.

Chaosed0 commented:
Appreciate your time testing the PR. In case it helps speed things along, I really don't care about credit for this - hhyyrylainen did most of the work anyway, and really I just want crash dumps :) so squash my commits as necessary.

huwpascoe commented:
@hhyyrylainen
4.3 is out, here's your chance.

hhyyrylainen commented:
Well as far as I know this is still good to go after being fixed in June. So if anyone with merge permissions would like to review this, that'd be pretty great. Or can someone tell me where I'd need to go to ask someone to review this?

This isn't yet super time critical for me as for my game it seems only Godot 4.4 will stop it from crashing 100% of the time, so currently for me these changes are not useful yet.

huwpascoe commented:
First it needs to be squashed, then sometime next week I'll start campaigning for this to be merged.

akien-mga commented:
It's not a requirement to squash before it has been reviewed. Once approved, it can be squashed, but there will likely be more work needed once someone does a deep dive to test and review.

Calinou reviewed (commented):
Tested locally, it works as expected.

Example crash dumps: [crashes.zip](https://github.com/user-attachments/files/16683748/crashes.zip)

Example output (I added `CRASH_NOW_MSG("Crash message example.")` to `project_manager.cpp` for testing):

```
ERROR: Crash message example.
   at: _notification (./editor/project_manager.cpp:103)
Crash dump created at: /home/hugo/.local/share/godot/app_userdata/[unnamed project]/crashes/bff06af4-d567-4e06-600635bf-35b8f3ae.dmp

================================================================
handle_crash: Program crashed with signal 4
Engine version: Godot Engine v4.4.dev.custom_build (c1f12813bf9d07cff283fae793aea385272640f9)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] /lib64/libc.so.6(+0x40d00) [0x7ff362c4fd00] (??:0)
[2] bin/godot.linuxbsd.editor.x86_64() [0x4cc0130] (/home/hugo/Documents/Git/godotengine/godot/./editor/project_manager.cpp:103 (discriminator 2))
[3] bin/godot.linuxbsd.editor.x86_64() [0x4cefd9f] (/home/hugo/Documents/Git/godotengine/godot/./editor/project_manager.h:56 (discriminator 14))
[4] bin/godot.linuxbsd.editor.x86_64() [0x9b46534] (/home/hugo/Documents/Git/godotengine/godot/./core/object/object.cpp:873)
[5] bin/godot.linuxbsd.editor.x86_64() [0x5c52cd5] (/home/hugo/Documents/Git/godotengine/godot/./scene/main/node.cpp:279)
[6] bin/godot.linuxbsd.editor.x86_64() [0x5c52c45] (/home/hugo/Documents/Git/godotengine/godot/./core/templates/hash_map.h:455)
[7] bin/godot.linuxbsd.editor.x86_64() [0x5c6b713] (/home/hugo/Documents/Git/godotengine/godot/./scene/main/node.cpp:3204)
[8] bin/godot.linuxbsd.editor.x86_64() [0x5d24c41] (/home/hugo/Documents/Git/godotengine/godot/./scene/main/scene_tree.cpp:473)
[9] bin/godot.linuxbsd.editor.x86_64() [0x2b1c28b] (/home/hugo/Documents/Git/godotengine/godot/./servers/display_server.h:60)
[10] bin/godot.linuxbsd.editor.x86_64() [0x2b11771] (/home/hugo/Documents/Git/godotengine/godot/platform/linuxbsd/godot_linuxbsd.cpp:85)
[11] /lib64/libc.so.6(+0x2a088) [0x7ff362c39088] (??:0)
[12] /lib64/libc.so.6(__libc_start_main+0x8b) [0x7ff362c3914b] (??:0)
[13] bin/godot.linuxbsd.editor.x86_64() [0x2b11565] (??:?)
-- END OF BACKTRACE --
================================================================
[1]    101724 IOT instruction (core dumped)  bin/godot.linuxbsd.editor.x86_64
```

Some comments:

- When the crash happens within the editor or project manager, the dump is saved in a project-specific folder. It should go to the editor data folder instead (`EditorPaths::get_data_dir()`).

- Could the dump filename be prefixed with an ISO 8601 timestamp? This would allow it to be sorted by date even if filesystem timestamps get overwritten for some reason (or when sharing the file to others).
  - For reference, we use `YYYY-MM-DDTHHMMSS` for editor screenshots (where `T` is a literal "T"). I suggest using the same format for consistency.

- I suggest rewording the message to make it easier to understand for people not familiar with Breakpad:
```
Breakpad crash dump created at: /path/to/crash.dmp
Please attach this file when reporting issues to the project developer.
```
  - Interestingly, GitHub doesn't allow uploading `.dmp` files in issue comments, even though the error message says it's one of the allowed extensions:
  > **We don’t support that file type.**
  > Try again with GIF, JPEG, JPG, MOV, MP4, PNG, SVG, WEBM, CPUPROFILE, CSV, DMP, DOCX, FODG, FODP, FODS, FODT, GZ, JSON, JSONC, LOG, MD, ODF, ODG, ODP, ODS, ODT, PATCH, PDF, PPTX, TGZ, TXT, XLS, XLSX or ZIP.

- `use_breakpad` is `False` by default, which means this feature is disabled by default. This might surprise people compiling from source, so maybe it should be enabled by default? I'm not sure, since official builds won't be able to benefit from it until [debugging symbols are distributed](https://github.com/godotengine/godot-proposals/issues/1342) or at least preserved somewhere.

- Is the `use_breakpad` SCons option needed when there's `module_breakpad_enabled`, which could default to `False`? We do something similar for C#, where `module_mono_enabled` is `False` by default.

Calinou commented on modules/breakpad/SCsub:
```diff
@@ -0,0 +1,187 @@
+
+thirdparty_dir = "#thirdparty/breakpad/"
+
+# TODO: find out when these are needed (if at all)
+dwarf_module = False
+stabs_module = False
```
Did anyone find that these were needed in the end? If not, I'd remove the relevant code (and perhaps keep a second branch on your fork with this code present).

hhyyrylainen replied:
If I remember right, no one has tried this out on mac yet. So there is a chance that these would be required on mac.

If it is determined necessary to get this PR merged, I now have a mac I can use for development so I can check things there to make sure the mac version compiles and can create crash dumps.

hhyyrylainen commented:
@Calinou thanks for the review. Most of that sounds doable.

> When the crash happens within the editor or project manager, the dump is saved in a project-specific folder. It should go to the editor data folder instead (EditorPaths::get_data_dir()).

Sounds easily doable. I'll make changes soon (I'm preparing a new release of my game this week which I has to get done this week so that takes priority).

> Could the dump filename be prefixed with an ISO 8601 timestamp? This would allow it to be sorted by date even if filesystem timestamps get overwritten for some reason (or when sharing the file to others).

I think this is going to be pretty difficult as I just tried looking and it doesn't seem like breakpad allows overwriting the file name. If I found the right place in the Breakpad code then a crash dump is always generated with a random GUID as the name. So it might be that the only way to update that part is to patch the library itself.

If I remember right, minidumps should have the crash time in the headers, so the actual crash time can be verified by reading the minidump file headers. And I did take a quick peek at the Breakpad code and it looks like it has some header update logic in the minidump writer class that updates the time: `      header.get()->time_date_stamp = time(NULL);`


> I suggest rewording the message to make it easier to understand for people not familiar with Breakpad:

I'll do this. Also if it isn't a bad idea, I'll add a tiny bit of delay like half a second (and trying to flush output) before quitting after printing that message. I'm hoping it helps with my usecase where a parent process reading the output for some reason never sees the crash dump info, though it does always appear in terminal. Also could slightly help with command prompt usage if it closes with a slight delay there's a chance the user kind of can see that something was printed.


> use_breakpad is False by default, which means this feature is disabled by default. This might surprise people compiling from source, so maybe it should be enabled by default? I'm not sure, since official builds won't be able to benefit from it until https://github.com/godotengine/godot-proposals/issues/1342 or at least preserved somewhere.

I guess that would be fine, but the official builds should not be compiled with Breakpad enabled until symbols can be distributed, because otherwise exported games will create crash dumps that no one will be able to decode. So I think those would cause more confusion than help with anything. If there's some way to default to on only when compiling outside the official release builds, I could do that.


> Is the use_breakpad SCons option needed when there's module_breakpad_enabled, which could default to False? We do something similar for C#, where module_mono_enabled is False by default.

I didn't actually know that there was a separate mechanism already to enable / disable parts of the engine. I think I saw some other feature also having a similar SCons option so I followed that, but I can switch to just having a `module_breakpad_enabled` as the toggle for this feature.

DanielSnd commented:
I merged this PR into my custom fork and it works well on my own tests here forcing crashes.

I looked into it, found it and merged it because a few of the players in one of my games are getting a very specific weird crash that I'm not able to reproduce. Weirdly they can reproduce it every time, and it's through the course of starting the tutorial level so not a great place to be crashing hahaha.

For some reason for this crash it doesn't trigger breakpad/the crash handler. It crashed straight out of the game with no dmp being created or no extra crash/error logs on the log file.

I'm not sure where to go from here, so figured I'd post here. Any ideas on why breakpad wouldn't trigger on a crash while it triggers when I'm deliberately trying to crash? And anything I can do to try and get it to trigger on this crash?

hhyyrylainen commented:
My immediate thought is that it may not be a "crash" but instead a forced shutdown of the game after detecting a problem. That would basically look like a crash to the user but not technically be one. Though I don't know enough to say for sure. At least back in the 3.x days I never received any word of a crash that didn't create a .dmp file (as far as I can recall).

Now that Godot 4.4 is out and my game doesn't crash 100% of the time on shutdown, I no longer have that blocker for reintroducing this into my game. But I don't know when I will next work on this.

illwieckz commented:
For reference since DæmonEngine's breakpad [was mentionned](https://github.com/godotengine/godot/pull/56014#issuecomment-997410856), we now merged Google upstream to our breakpad fork maintaining MinGW support. So it is now not only being on par with Google implementation, but also still provides the precious MinGW support.

The big highlight is that this now supports DWARF5 symbols in both Linux and MinGW executables:

- https://github.com/DaemonEngine/breakpad
